### PR TITLE
Implement `InferenceClient.chat_completion` + use new types for text-generation

### PR DESCRIPTION
**Pull request by Wauplin**
Solve #2064.

Includes:
- types for text-generation + chat-completion outputs  (related to https://github.com/huggingface/huggingface.js/pull/537)
- support for the Chat completion in both `InferenceClient` and `AsyncInferenceClient`
- use text-generation types, including when stream=True
- make `minijinja` an optional dependency => only needed when models are not TGI-powered

### Workflow:

The workflow is a bit hacky depending on how the model is served. If the model/url is compatible with the Message API (which is the case for TGI-powered models), it's fairly straightforward. Otherwise we need to fetch the chat_completion config from the Hub, render the prompt and send a `text-generation` request. Rendering in done by [minijinja](https://github.com/mitsuhiko/minijinja) a lightweight rust-based template engine.

TL;DR: Works fine in most cases except for models running on transformers and served in Inference Endpoint. UX is quite bad for them (requests needs to be done by the user owning the inference endpoint). 

chat_completion

### TODO:
- [x] tests
- [x] documentation
- [x] handle TGI-served models
- [x] handle from TGI-based Inference Endpoint URL
- [x] handle from transformers-based Inference Endpoint URL

### Example (stream=True)

```py
from huggingface_hub import InferenceClient

client = InferenceClient("HuggingFaceH4/zephyr-7b-beta")

messages = [
    {"role": "system", "content": "You are a cook chef and always provide very good cooking advices."},
    {"role": "user", "content": "How can I cook spinach? List me 5 recipes with a very short description for each of them. I want diverse recipes from around the world. The spinach must be the main ingredient of the dish."},
]

# Print messages
for message in messages:
    print(message["role"].upper())
    print(message["content"].strip())
    print()

# Send request + print response
print("ASSISTANT")
response = ""
for item in client.chat_completion(messages=messages, max_tokens=10000, stream=True):
    if item.choices[0].finish_reason is None:
        delta_content = item.choices[0].delta.content
        response += delta_content
        print(delta_content, end="")
print("\n")
messages.append({"role": "assistant", "content": response})

# New user message
messages.append({"role": "user", "content": "Thank you! Could you give me the recipe for the third one with ingredient list and preparation steps?"})
print(messages[-1]["role"].upper())
print(messages[-1]["content"].strip())
print()

# Send new request + print response
print("ASSISTANT")
response = ""
for item in client.chat_completion(messages=messages, max_tokens=10000, stream=True):
    if item.choices[0].finish_reason is None:
        delta_content = item.choices[0].delta.content
        response += delta_content
        print(delta_content, end="")
print()
messages.append({"role": "assistant", "content": response})
```

### Output

```
SYSTEM
You are a cook chef and always provide very good cooking advices.

USER
How can I cook spinach? List me 5 recipes with a very short description for each of them. I want diverse recipes from around the world. The spinach must be the main ingredient of the dish.

ASSISTANT
1. Indian Palak Paneer: This vegetarian dish is a classic from North India. Sauté spinach with onions, garlic, and spices until wilted, then add crumbled paneer (Indian cheese) and simmer until heated through. Serve with rice or naan bread.

2. Greek Spanakorizo: This vegetarian dish is a traditional Greek rice dish made with spinach, rice, onions, garlic, and spices. It's often served as a side dish or as a vegetarian main course.

3. Italian Spinach and Ricotta Stuffed Shells: Cook jumbo pasta shells, then stuff them with a mixture of spinach, ricotta cheese, and spices. Bake in a tomato sauce until bubbly and golden brown.

4. Filipino Sinigang na Spinach: This sour soup is a staple in Filipino cuisine. Spinach is cooked with tomatoes, onions, garlic, and tamarind paste until tender, then seasoned with fish sauce and vinegar. Serve with rice.

5. Middle Eastern Spinach and Feta Stuffed Pita Bread: Cut pita bread in half and stuff with a mixture of spinach, feta cheese, onions, and spices. Bake in the oven until the bread is crispy and the filling is heated through. Serve with a side salad.

USER
Thank you! Could you give me the recipe for the third one with ingredient list and preparation steps?

ASSISTANT
Sure, here's the recipe for Spinach and Ricotta Stuffed Shells:

Ingredients:
- 12 jumbo pasta shells
- 1 (10-ounce) package frozen chopped spinach, thawed and drained
- 1 (15-ounce) container ricotta cheese
- 1/2 cup grated Parmesan cheese
- 1/2 cup grated mozzarella cheese
- 1/4 teaspoon salt
- 1/4 teaspoon black pepper
- 2 cloves garlic, minced
- 24-ounce jar marinara sauce
- 1/4 cup chopped fresh parsley

Preparation:

1. Preheat the oven to 375°F. Cook the pasta shells according to package directions. Drain and rinse with cold water to stop the cooking process.

2. In a large bowl, combine the spinach, ricotta cheese, Parmesan cheese, mozzarella cheese, salt, pepper, and garlic. Mix well.

3. Spoon the spinach mixture into the cooked pasta shells, filling each shell generously.

4. Pour the marinara sauce into the bottom of a 9x13-inch baking dish. Arrange the stuffed shells in the dish, placing them in a single layer.

5. Sprinkle chopped parsley over the top of the stuffed shells.

6. Bake for 25-30 minutes, or until the shells are heated through and the cheese is melted and bubbly.

7. Let the stuffed shells cool for a few minutes before serving. Enjoy!
```

### CODE REVIEW

File: src/huggingface_hub/inference/_client.py [hunk 1]
```diff
@@ -395,6 +400,100 @@ def automatic_speech_recognition(
...
+        >>> client.chat_completion(...)
+        """
+        # determine model
+        model = model or self.model or self.get_recommended_model("text-generation")
+
+        # TODO: add support when model is a URL
+        assert not model.startswith("https://") and not model.startswith("http://")
+
+        # fetch chat template + tokens
+        prompt = render_chat_prompt(model_id=model, token=self.token, messages=messages)
```
[the remark this answers is not in the document]
Comment (Wauplin) [reply src/huggingface_hub/inference/_client.py:465]: Yes actually that's what I want to do as well! Both for urls and model ids. We already have a `_is_tgi_server` helper that can be reused (same logic as the JS one)

And I also need to handle streaming properly.

[the remark this answers is not in the document]
Comment (SBrandeis) [reply src/huggingface_hub/inference/_client.py:465]: > if model is hosted in TGI, POST'ing the messages directly

`ConversationalWidget` in huggingface.js does not use the chat completion API (yet) - instead, it uses jinja to format the input with the chat template and sends that to the text generation API

[the remark this answers is not in the document]
Comment (SBrandeis) [reply src/huggingface_hub/inference/_client.py:465]: TGI conditions the use of streaming tho

[the remark this answers is not in the document]
Comment (julien-c) [reply src/huggingface_hub/inference/_client.py:465]: Ok, it'd be a nice improvement i think (not really for performance as I expect the perf to be the same, but for "elegance")

**Comment by LysandreJik** [comment]
cc @Rocketknight1 as well as it'd be great if this + the conversational implem in transformers were very similar

File: src/huggingface_hub/inference/_client.py [hunk 2]
```diff
@@ -395,6 +409,262 @@ def automatic_speech_recognition(
...
+            temperature=temperature,
+            top_p=top_p,
+        )
+
+        created = int(time.time())
+
+        if stream:
+            return _stream_chat_completion_response_from_text_generation(text_generation_output)  # type: ignore [arg-type]
+
+        if isinstance(text_generation_output, TextGenerationOutput):
```
Comment (SBrandeis) [inline src/huggingface_hub/inference/_client.py:640]: In which case is the output not a `TextGenerationOutput` instance?

File: src/huggingface_hub/inference/_client.py [hunk 3]
```diff
@@ -395,6 +409,262 @@ def automatic_speech_recognition(
...
+                        index=0,
+                        message=ChatCompletionOutputChoiceMessage(
+                            content=text_generation_output,  # type: ignore[arg-type]
+                            role="assistant",
+                        ),
+                    )
+                ],
+            )
+
     def conversational(
```
Comment (SBrandeis) [inline src/huggingface_hub/inference/_client.py:668]: Should be marked as `@deprecated` ?

**Review by SBrandeis — APPROVED** [review]
nice

File: src/huggingface_hub/inference/_client.py (hunk 2 above)
Comment (Wauplin) [reply src/huggingface_hub/inference/_client.py:640]: If server doesn't support `details=True` (typically if served with transformers instead of TGI), then it returns a simple string. That's why I need to format the output differently. I'll add a comment :+1:

File: pyproject.toml [hunk 4]
```diff
@@ -1,9 +1,6 @@
 [tool.mypy]
 ignore_missing_imports = true
 no_implicit_optional = true
-plugins = [
-    "pydantic.mypy",
-]
```
Comment (LysandreJik) [inline pyproject.toml:6]: :eyes:

File: src/huggingface_hub/inference/_client.py [hunk 5]
```diff
@@ -143,12 +147,13 @@ class InferenceClient:
     def __init__(
         self,
         model: Optional[str] = None,
-        token: Union[str, bool, None] = None,
+        token: Optional[str] = None,
```
Comment (LysandreJik) [inline src/huggingface_hub/inference/_client.py:150]: If this is a voluntary change, you should update the docstring above as well

Comment (LysandreJik) [reply src/huggingface_hub/inference/_client.py:150]: (passing `False` would be unexpected

File: src/huggingface_hub/inference/_client.py [hunk 6]
```diff
@@ -395,6 +409,258 @@ def automatic_speech_recognition(
...
+        messages: List[Dict[str, str]],
+        *,
+        model: Optional[str] = None,
+        stream: bool = False,
+        max_tokens: int = 20,
+        seed: Optional[int] = None,
+        stop: Optional[Union[List[str], str]] = None,
+        temperature: float = 1.0,
+        top_p: Optional[float] = None,
+    ) -> Union[ChatCompletionOutput, Iterable[ChatCompletionStreamOutput]]:
```
Comment (LysandreJik) [inline src/huggingface_hub/inference/_client.py:465]: Is it expected that the overload needs to be defined with the exact same current signature for the method? (these two seem to be exactly the same)

File: src/huggingface_hub/inference/_client.py [hunk 7]
```diff
@@ -395,6 +409,258 @@ def automatic_speech_recognition(
...
+            - if `stream=False`, the generated text is returned as a [`ChatCompletionOutput`] (default).
+            - if `stream=True`, the generated text is returned token by token as a sequence of [`ChatCompletionStreamOutput`].
+
+        Raises:
+            [`InferenceTimeoutError`]:
+                If the model is unavailable or the request times out.
+            `HTTPError`:
+                If the request fails with an HTTP error status code other than HTTP 503.
+
+        Example:
```
Comment (LysandreJik) [inline src/huggingface_hub/inference/_client.py:516]: Awesome example!

File: src/huggingface_hub/inference/_client.py [hunk 8]
```diff
@@ -395,6 +409,258 @@ def automatic_speech_recognition(
...
+        to install it.
+
+        </Tip>
+
+        Args:
+            messages (List[Union[`SystemMessage`, `UserMessage`, `AssistantMessage`]]):
+                Conversation history consisting of roles and content pairs.
+            model (Optional[`str`], optional):
+                Unique identifier of the model to use for inference.
+                Defaults to None.
```
Comment (LysandreJik) [inline src/huggingface_hub/inference/_client.py:484]: Maybe mention what happens when defaulting to `None`? Following your graph it should be zephyr-7b-beta

Comment (LysandreJik) [reply src/huggingface_hub/inference/_client.py:484]: Or maybe link to where this data got from so that they can check the current default model? https://github.com/huggingface/huggingface_hub/blob/dbf8c1018dd91731b3a3286879d99a8d4ec3f81b/src/huggingface_hub/inference/_common.py#L162-L164

File: src/huggingface_hub/inference/_client.py [hunk 9]
```diff
@@ -395,6 +409,258 @@ def automatic_speech_recognition(
...
+        if _is_chat_completion_server(model):
+            # First, let's consider the server has a `/v1/chat/completions` endpoint.
+            # If that's the case, we don't have to render the chat template client-side.
+            model_url = self._resolve_url(model) + "/v1/chat/completions"
+
+            try:
+                data = self.post(
+                    model=model_url,
+                    json=dict(
+                        model="tgi",  # random string
```
Comment (LysandreJik) [inline src/huggingface_hub/inference/_client.py:556]: Is this a required key even though `model` is already defined above?

File: src/huggingface_hub/inference/_client.py [hunk 10]
```diff
@@ -395,6 +409,258 @@ def automatic_speech_recognition(
...
+            # A workaround is to list the user Inference Endpoints and check if one of them correspond to the model URL.
+            # If not, we raise an error.
+            # TODO: fix when we have a proper API for this (at least for Inference Endpoints)
+            # TODO: what if Sagemaker URL?
+            # TODO: what if Azure URL?
+            from ..hf_api import HfApi
+
+            for endpoint in HfApi(token=self.token).list_inference_endpoints():
+                if endpoint.url == model:
+                    model_id = endpoint.repository
```
Comment (LysandreJik) [inline src/huggingface_hub/inference/_client.py:604]: I had in mind that inference endpoints could also serve models TGI; in this case shouldn't the server-based chat_completion work with that model?

File: src/huggingface_hub/inference/_client.py [hunk 11]
```diff
@@ -446,6 +719,11 @@ def conversational(
         ... )
         ```
         """
+        warnings.warn(
+            "'InferenceClient.conversational' is deprecated and will be removed starting from huggingface_hub>=0.25. "
+            "Please you the more appropriate 'InferenceClient.chat_completion' API instead.",
```
Comment (LysandreJik) [inline src/huggingface_hub/inference/_client.py:724]:
```suggestion
            "Please use the more appropriate 'InferenceClient.chat_completion' API instead.",
```

File: src/huggingface_hub/inference/_templating.py [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+    minijinja = _import_minijinja()
+    template = _fetch_and_compile_template(model_id=model_id, token=token)
+
+    try:
+        return template(messages=messages, add_generation_prompt=add_generation_prompt, **kwargs)
+    except minijinja.TemplateError as e:
+        raise TemplateError(f"Error while trying to render chat prompt for model '{model_id}': {e}") from e
+
+
+@lru_cache  # TODO: lru_cache for raised exceptions
```
Comment (LysandreJik) [inline src/huggingface_hub/inference/_templating.py:52]: Does that mean that if a remote repository updates its chat template, then a runtime restart will be necessary to get the updated template?

Comment (LysandreJik) [reply src/huggingface_hub/inference/_templating.py:52]: (probably nitpick)

**Review by LysandreJik — APPROVED** [review]
Very impressive PR, and a significant improvement to what existed before! I've added a few comments but I'm approving the overall logic/implementation so as to not block you to merge + release.

Let me know if you'd like another review once the comments are addressed

File: pyproject.toml (hunk 4 above)
Comment (Wauplin) [reply pyproject.toml:6]: Not needed anymore so prefer to remove it (we can always re-add it if we introduce pydantic again in the future)

File: src/huggingface_hub/inference/_client.py (hunk 5 above)
Comment (Wauplin) [reply src/huggingface_hub/inference/_client.py]: Updated to revert to `Union[str, bool, None]` + fix related issue + updated docstrings in https://github.com/huggingface/huggingface_hub/pull/2094/commits/11c37a5a45e8b3b52e5126d5e7754a28b0759c39. It's more of a documentation / typing issue since bool value is always correctly respected (but in some cases I don't want to document it since it's not relevant when auth is required)

File: src/huggingface_hub/inference/_client.py (hunk 9 above)
Comment (Wauplin) [reply src/huggingface_hub/inference/_client.py:556]: Yes, we are mimic-ing the "official" API and therefore some args are useless but we still need to pass it. TGI will not use this value.

File: src/huggingface_hub/inference/_client.py (hunk 10 above)
Comment (Wauplin) [reply src/huggingface_hub/inference/_client.py:604]: Yes and when that's the case (e.g. for 90%+ of LLMs) the condition above "`if _is_chat_completion_server(model):`"  is True, so we never execute this part of the logic. This is just a "let's default back to a hacky solution but hoping we don't have to rely on it". (it works fine though, tested and approved :) )

File: src/huggingface_hub/inference/_templating.py (hunk 12 above)
Comment (Wauplin) [reply src/huggingface_hub/inference/_templating.py:52]: > Does that mean that if a remote repository updates its chat template, then a runtime restart will be necessary to get the updated template?

Yes indeed, that's the case.
Not worth improving IMO. And in any case, this is "only" for models not served with TGI so not the common use case.

File: src/huggingface_hub/inference/_client.py (hunk 8 above)
Comment (Wauplin) [reply src/huggingface_hub/inference/_client.py]: Nice catch. Updated in https://github.com/huggingface/huggingface_hub/pull/2094/commits/9aa581c6a55464e8a0b2a01efa3e3182f185e3e9.

File: src/huggingface_hub/inference/_client.py (hunk 6 above)
Comment (Wauplin) [reply src/huggingface_hub/inference/_client.py:465]: So... I had to fight a bit with mypy and this is the only solution I've found to make it right :grin:

I _think_ once you've started using @overload, the default definition with no @overload is not taken into account, hence the need for a redefinition. As of why I need to define it it's because mypy doesn't make the relation that `bool` is equivalent to `Union[Literal[True], Literal[False]]` (defined above). Anyway, it's a bit ugly like this but if it can help users to expect the correct type, that's for the best.

**Comment by Wauplin** [comment]
Thanks a lot for the thorough reviews @SBrandeis @LysandreJik! I think I've addressed all the comments so I'll soon merge it (waiting for the CI to complete). I'm glad to have this topic closed :smile:

**Comment by OlivierDehaene** [comment]
Coming back to this, would having a `api-inference/tokenizer/<model_id>/apply_chat_template` POST route for all models be something that would help you?

**Comment by Wauplin** [comment]
Hey @OlivierDehaene, thanks for the proposition! Actually @LysandreJik just told me that transformers-served LLMs can now receive a list of messages instead of an input text. This is already available out-of-the-box in InferenceAPI so I don't even need to render the template client-side. H/t to @Rocketknight1 who implemented it.

So a POST route `api-inference/tokenizer/<model_id>/apply_chat_template` is not really needed anymore -at least not for my use case.
